### PR TITLE
Omit empty kernel/firmware directories from initrd

### PR DESCRIPTION
**Pull request by septatrix**
Extracted into a separate PR as suggested in https://github.com/systemd/mkosi/pull/2521#pullrequestreview-1941599887

While the space savings are not huge they are measurable, especially when having a lot of modules in the main image and very few in the initrd. In my case I have the `@hardware-support` fedora group installed which pulls in pretty much every driver though I have excluded almost all of those from the initrd.

Additionally, this makes the directory structure a lot cleaner and easier to inspect which kernel modules and firmware files are actually included.

### CODE REVIEW

File: mkosi/kmod.py [hunk 1]
```diff
@@ -172,30 +172,22 @@ def gen_required_kernel_modules(
...
-        yield p
-
-    for p in (root / modulesd).iterdir():
-        if p.name.startswith("modules"):
-            yield p
+    # We do not want to yield directories lower than /usr/lib, especially none outside of our root namespace.
+    # Therefore we disregard the lowest few parent directores.
+    # This also improves performance as fewer set memberships have to be checked.
+    prefix_length = len((root / "usr/lib").parts)
+    required_paths = mods | firmware | set(
```
Comment (keszybz) [inline mkosi/kmod.py:179]: Please just use the set comprehension: `{ parent for … }`

**Review by keszybz — COMMENTED** [review]
Looks reasonable.

File: mkosi/kmod.py (hunk 1 above)
Comment (behrmann) [reply mkosi/kmod.py:179]: The whole thing could also be written as
```python
firmwared = root / "usr/lib/firmware"
required_paths = mods | firmware | { 
    parent
    for file in mods | firmware
    for parent in file.parents
    if parent.is_relative_to(modulesd) or parent.is_relative_to(firmwared)
}
```
It's a bit clearer (no list index juggling) and performs the same. It's also a bit tighter in what it accepts.

**Review by behrmann — APPROVED** [review]
LGTM, same nit as @keszybz and another suggestion.

**Comment by DaanDeMeyer** [comment]
I've just put up https://github.com/systemd/mkosi/pull/2523 which fixes this in a (subjectively) cleaner way

**Comment by DaanDeMeyer** [comment]
Let's close this as the other PR was merged

**Pull request closed without merging**